README.md
# GolGrid

Conway's [Game of Life] in a flutter widget.

A [`GolGrid`] is a rendering of a [`GridWorld`] controlled by a [`Thumper`].

### Example

The [example] contains two alternatives.

One uses several relatively small widgets in a scrollable column,
suitable for a browser:

[![movie web](https://img.youtube.com/vi/DIPRQGnh4Nc/0.jpg)](https://youtu.be/DIPRQGnh4Nc)

The other does a media query to do a full
screen demo with a complex world (four [Gosper guns]
and an [R-pentomino]), more suitable for android:

[![movie android][android1]](https://youtu.be/x6oBF_giM-s)

To run the examples, install [flutter] (for `demo-web`
demo, use [beta channel flutter]), then

```bash
# Confirm flutter installed and devices available.
flutter devices

# Get the code
git clone git@github.com:monopole/gol_grid.git
cd gol_grid

# then either
make demo-android # needs USB connected device
# or
make demo-web # needs chrome
```


[beta channel flutter]: https://flutter.dev/docs/get-started/web
[`GolGrid`]: https://pub.dev/packages/gol_grid
[`GridWorld`]: https://pub.dev/packages/grid_world
[`Thumper`]: https://pub.dev/packages/thumper
[Game of Life]: https://en.wikipedia.org/wiki/Conway%27s_Game_of_Life
[Gosper guns]: https://en.wikipedia.org/wiki/Gun_(cellular_automaton)
[R-pentomino]: https://en.wikipedia.org/wiki/Conway%27s_Game_of_Life#Examples_of_patterns
[spaceship]: https://en.wikipedia.org/wiki/Conway%27s_Game_of_Life#Examples_of_patterns
[example]: example/lib/main.dart
[android1]: images/android1.png
[inVideo]: https://youtu.be/DIPRQGnh4Nc
[inMovie]: images/demo-chrome-at-v0.1.7.mp4
[flutter]: https://flutter.dev/docs/get-started/install


CHANGELOG.md
## 0.2.0
 - upgrade to bloc 8.0.1

## 0.1.8
 - android movie

## 0.1.7
 - improve chrome demo.

## 0.1.6
 - Depend on thumper 2.0.0

## 0.1.5
 - Depend on thumper 1.0.0, grid_world 1.0.1

## 0.1.4
 - Upgrade to latest thumper, add screenshots.

## 0.1.3
 - Tweak example.

## 0.1.2
 - Tweak example.

## 0.1.1
 - Tweak example.

## 0.1.0
 - Initial package commit.


test/gol_grid_test.dart
import 'package:gol_grid/gol_grid.dart';
import 'package:grid_world/grid_world.dart';
import 'package:test/test.dart';

void main() {
  test('toString', () {
    final grid = GolGrid(DimensionedWorld.make(ConwayEvolver.glider));
    expect(grid.toString(), equals('GolGrid'));
  });

  // testWidgets('TODO add a bloc and test', (tester) async {
  //   // Build our app and trigger a frame.
  //   await tester.pumpWidget(GolGrid(GolGridDimensions()));
  //
  //   // Verify that our counter starts at 0.
  //   expect(find.text('0'), findsOneWidget);
  //   expect(find.text('1'), findsNothing);
  //
  //   // Tap the '+' icon and trigger a frame.
  //   await tester.tap(find.byIcon(Icons.add));
  //   await tester.pump();
  //
  //   // Verify that our counter has incremented.
  //   expect(find.text('0'), findsNothing);
  //   expect(find.text('1'), findsOneWidget);
  // });
}


test/dimensioned_world_test.dart
import 'package:test/test.dart';
import 'package:flutter/material.dart';
import 'package:gol_grid/src/dimensioned_world.dart';
import 'package:grid_world/grid_world.dart';

void main() {
  late GridWorld gw;
  late DimensionedWorld dw;

  setUp(() {
    gw = ConwayEvolver.blinker.padBottom(1);
    dw = DimensionedWorld.make(gw, lineWidth: 3, cellWidth: 7);
  });

  test('ctor', () {
    expect(dw.gridWorld, gw);
    expect(dw.gridWorld.toString(), '''
.....
..#..
..#..
..#..
.....
.....
''');
    expect(dw.cellSize, const Size(7, 7));
    expect(dw.gridWorld.nCols, 5);
    expect(dw.gridWorld.nRows, 6);
    //  width: (nCols*(7+3))-3 = 47
    // height: (nRows*(7+3))-3 = 57
    expect(dw.worldSize, const Size(47, 57));
    expect(dw.unusedWidth, 0);
    expect(dw.unusedHeight, 0);
    expect(dw.offset(0), 0);
    expect(dw.offset(1), 10);
    expect(dw.offset(2), 20);
  });

  test('expandToFit', () {
    dw = dw.expandToFit(60, 80);
    expect(dw.gridWorld.toString(), '''
......
......
..#...
..#...
..#...
......
......
......
''');
    expect(dw.cellSize, const Size(7, 7));
    expect(dw.gridWorld.nCols, 6);
    expect(dw.gridWorld.nRows, 8);
    //  width: (nCols*(7+3))-3 = 57
    // height: (nRows*(7+3))-3 = 77
    expect(dw.worldSize, const Size(57, 77));
    // 60-57 = 3
    expect(dw.unusedWidth, 3);
    // 80-77 = 3
    expect(dw.unusedHeight, 3);
    expect(dw.offset(0), 0);
    expect(dw.offset(1), 10);
    expect(dw.offset(2), 20);
  });
}


lib/src/gol_grid.dart
import 'package:flutter/material.dart';
import 'package:flutter_bloc/flutter_bloc.dart';
import 'package:grid_world/grid_world.dart';
import 'package:thumper/thumper.dart';
import 'dimensioned_world.dart';

// ignore_for_file: diagnostic_describe_all_properties

/// GolGrid is a widget showing a GridWorld controlled by a Thumper<GridWorld>.
@immutable
class GolGrid extends StatelessWidget {
  /// Returns a widget sized to its [DimensionedWorld] argument.
  /// It's assumed that the worlds provided in the state stream from
  /// ThumperBloc<GridWorld> will all have matching dimensions (this is
  /// assured if they came from a [GridWorldIterable]).
  const GolGrid(
    this._dimensions, {
    Key? key,
    this.foregroundColor = Colors.lightBlueAccent,
    this.backgroundColor = Colors.black,
    this.controlsForegroundColor = Colors.lightGreenAccent,
    this.controlsBackgroundColor = Colors.black,
    this.hardIterationLimit = 10000,
  }) : super(key: key);

  /// [GridWorld] plus dimensions associated with rendering.
  final DimensionedWorld _dimensions;

  /// Game of life live cell color.
  final Color foregroundColor;

  /// Game of life dead cell color.
  final Color backgroundColor;

  /// Thumper foreground color.
  final Color controlsForegroundColor;

  /// Thumper background color.
  final Color controlsBackgroundColor;

  /// When this limit is hit, automatically halt iteration.
  /// No way to continue - must reset and start over.
  /// This is nothing more than energy conservation.
  final int hardIterationLimit;

  @override
  Widget build(BuildContext context) => BlocProvider(
        create: (context) => ThumperBloc<GridWorld>.fromIterable(
            GridWorldIterable(_dimensions.gridWorld,
                limit: hardIterationLimit)),
        child: BlocBuilder<ThumperBloc<GridWorld>, ThumperState<GridWorld>>(
          buildWhen: (previousState, incomingState) =>
              incomingState.thumpCount != previousState.thumpCount,
          builder: (ctx, state) => _column(state.thing),
        ),
      );

  Widget _column(GridWorld gw) => Container(
        width: _dimensions.worldSize.width + _dimensions.unusedWidth,
        height: _dimensions.worldSize.height +
            _dimensions.unusedHeight +
            Thumper.height,
        color: backgroundColor,
        padding: EdgeInsets.only(
            left: _dimensions.unusedWidth / 2,
            right: _dimensions.unusedWidth / 2),
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            CustomPaint(
              foregroundPainter: _CellPainter(gw, _dimensions, foregroundColor),
              //painter:
              //    _BackgroundPainter(_dimensions.worldSize, backgroundColor),
              size: _dimensions.worldSize,
            ),
            SizedBox(height: _dimensions.unusedHeight),
            Container(
                color: controlsBackgroundColor,
                child: Thumper<GridWorld>(onColor: controlsForegroundColor)),
          ],
        ),
      );
}

// _BackgroundPainter paints what's behind the cells.
// Could use this to, say, draw fancy throbbing grid lines,
// recently dead cell animations, etc.
// Not using this because if the background is just a solid color,
// the widget can just rely on the encapsulating [Container]'s color.
// The [Container] color should match the color of a dead cell,
// or we'll be obligated to explicitly draw dead cells, which
// doesn't happen at the time of writing.
// ignore: unused_element
class _BackgroundPainter extends CustomPainter {
  _BackgroundPainter(Size s, Color c)
      : _rect = Rect.fromLTRB(0, 0, s.width, s.height),
        _paint = _makeBackgroundPaint(c);

  final Rect _rect;
  final Paint _paint;

  static Paint _makeBackgroundPaint(Color color) => Paint()..color = color;

  @override
  void paint(Canvas canvas, Size size) {
    canvas.drawRect(_rect, _paint);
  }

  /// The painter's properties don't change.
  @override
  bool shouldRepaint(_BackgroundPainter oldPainter) => false;
}

/// _CellPainter knows how wide cells are, their color,
/// their separation, etc.  It doesn't draw grid lines
/// or dead cells; that's just the background
/// showing through.
class _CellPainter extends CustomPainter {
  _CellPainter(this._gw, this._dw, Color foregroundColor)
      : _cellPaint = _makeCellPaint(foregroundColor);

  // Used to set dimensions.
  final DimensionedWorld _dw;
  // Has changing cells, but fixed dimensions.
  final GridWorld _gw;
  final Paint _cellPaint;

  /// Cells could presumably be partially transparent too.
  static Paint _makeCellPaint(Color color) => Paint()..color = color;

  @override
  void paint(Canvas canvas, Size size) {
    for (var i = 0; i < _gw.nRows; i++) {
      for (var j = 0; j < _gw.nCols; j++) {
        if (_gw.isAlive(i, j)) {
          // Only draw live cells; assume dead cell matches background.
          final r = Offset(_dw.offset(j), _dw.offset(i)) & _dw.cellSize;
          canvas.drawRect(r, _cellPaint);
        }
      }
    }
  }

  /// The painter's properties don't change.
  @override
  bool shouldRepaint(_CellPainter oldPainter) => false;
}


lib/src/dimensioned_world.dart
import 'package:flutter/material.dart';
import 'package:grid_world/grid_world.dart';

/// [DimensionedWorld] holds lengths in units logical pixels associated
/// with rendering a [GridWorld].
///
/// E.g. the size of a cell, the distance between cells in pixels,
/// the size of the entire world rendered in pixels.
///
/// [DimensionedWorld] stores the [GridWorld] too, making it possible
/// to offer the [DimensionedWorld.expandToFit] method.
@immutable
class DimensionedWorld {
  /// Make a DimensionedWorld.
  factory DimensionedWorld.make(GridWorld w,
      {int lineWidth = 1, int cellWidth = 3}) {
    assert(cellWidth > 0, 'cellWidth must be > zero');
    assert(lineWidth > 0, 'lineWidth must be > zero');
    final cellSize = _makeCellSize(cellWidth);
    // worldSize gets used a lot, so caching it as a precomputed Size.
    final worldSize = _makeWorldSize(w, lineWidth, cellWidth);
    return DimensionedWorld._(
      w,
      lineWidth,
      cellWidth,
      cellSize,
      worldSize,
    );
  }

  /// Private constructor that accepts vetted arguments for caching.
  const DimensionedWorld._(this.gridWorld, this._lineWidth, this._cellWidth,
      this.cellSize, this.worldSize,
      {this.unusedWidth = 0, this.unusedHeight = 0})
      : assert(unusedWidth >= 0, 'unusedWidth must be >= 0'),
        assert(unusedHeight >= 0, 'unusedWidth must be >= 0');

  /// The world to be rendered.
  final GridWorld gridWorld;

  /// Grid line width in logical pixels.
  final int _lineWidth;

  /// Grid cell width in logical pixels.
  final int _cellWidth;

  /// Cell size in logical pixels. Cells are square.
  final Size cellSize;

  /// World size in logical pixels. Worlds are rectangular.
  final Size worldSize;

  /// Width left over from [expandToFit].
  final double unusedWidth;

  /// Height left over from [expandToFit].
  final double unusedHeight;

  /// Cells are square.
  static Size _makeCellSize(int w) => Size(w.toDouble(), w.toDouble());

  /// Worlds are rectangular.
  /// This offers width and height of a grid of cells
  /// with no lines on the borders, only in between cells.
  static Size _makeWorldSize(GridWorld gw, int lW, int cW) =>
      Size(_offset(gw.nCols, lW, cW) - lW, _offset(gw.nRows, lW, cW) - lW);

  /// Where is the left, top corner of (square) cell number i?
  double offset(int i) => _offset(i, _lineWidth, _cellWidth);

  /// This means no grid lines on borders.
  static double _offset(int i, int lW, int cW) => (i * (cW + lW)).toDouble();

  /// How many (discrete, complete) cells can be crammed into the given length?
  /// Takes the line (spacer) between each cell into account.
  ///
  /// Two ways to do this.
  ///
  /// With grid lines on all sides of the world:
  ///
  ///  length = ((nCells + 1)*_lineWidth) + (nCells * _cellWidth)
  ///         = (nCells * _lineWidth) + _lineWidth + (nCells * _cellWidth)
  ///  qed: nCells = (length - _lineWidth) / (_lineWidth + _cellWidth)
  ///
  /// With no grid lines on the sides:
  ///
  ///  length = ((nCells - 1)*_lineWidth) + (nCells * _cellWidth)
  ///         = (nCells * (_lineWidth + _cellWidth)) - _lineWidth
  ///  qed: nCells = (length + _lineWidth) / (_lineWidth + _cellWidth)
  ///
  /// Must take the floor() since we cannot allow fractional cells.
  int _howManyCellsFit(double length) =>
      ((length + _lineWidth) / (_lineWidth + _cellWidth)).floor();

  /// Add rows and columns to the world, to give the automata
  /// as much space as possible to maneuver.
  /// This throws if the underlying [GridWorld] is already too
  /// big to fit into the new width and height.
  /// Could try using ClipRect over an oversized GridWorld.
  DimensionedWorld expandToFit(double width, double height) {
    final gw = gridWorld.expandToFit(
        _howManyCellsFit(width), _howManyCellsFit(height));
    final wSize = _makeWorldSize(gw, _lineWidth, _cellWidth);
    return DimensionedWorld._(gw, _lineWidth, _cellWidth, cellSize, wSize,
        unusedWidth: width - wSize.width, unusedHeight: height - wSize.height);
  }
}


lib/src/gol_grid_dimensions.dart
import 'package:flutter/material.dart';
import 'package:grid_world/grid_world.dart';
import 'package:thumper/thumper.dart';

/// Calculates various dimensions.
@immutable
class GolGridDimensions {
  /// Make the dimensions.
  GolGridDimensions({this.lineWidth = 1, this.cellWidth = 3})
      : assert(cellWidth > 0 && lineWidth > 0, 'nonsense values'),
        cellSize = _makeCellSize(cellWidth);

  /// Grid line width in logical pixels.
  final int lineWidth;

  /// Grid cell width in logical pixels.
  final int cellWidth;

  /// Cell size in logical pixels. Cells are square.
  final Size cellSize;

  /// Cells are square.
  static Size _makeCellSize(int w) => Size(w.toDouble(), w.toDouble());

  /// Lets the client see how many Gol cells will fit in a given height, given
  /// the fixed cell, line and control widths used by this widget.
  int cellCountInHeight(double length) =>
      howManyCellsFit(length - Thumper.height);

  /// Like [cellCountInHeight], but for width.
  int cellCountInWidth(double length) => howManyCellsFit(length);

  /// How many cells can be crammed into the given length?
  /// Takes the line (spacer) between each cell into account.
  ///
  /// Two ways to do this.
  ///
  /// With grid lines on either side of the world:
  ///
  ///  length = ((nCells + 1)*_lineWidth) + (nCells * _cellWidth)
  ///         = (nCells * _lineWidth) + _lineWidth + (nCells * _cellWidth)
  ///  qed: nCells = (length - _lineWidth) / (_lineWidth + _cellWidth)
  ///
  /// No grid lines on either side:
  ///
  ///  length = ((nCells - 1)*_lineWidth) + (nCells * _cellWidth)
  ///         = (nCells * (_lineWidth + _cellWidth)) - _lineWidth
  ///  qed: nCells = (length + _lineWidth) / (_lineWidth + _cellWidth)
  ///
  /// In other case, we have to take the floor.
  int howManyCellsFit(double length) =>
      ((length + lineWidth) / (cellWidth + lineWidth)).floor();

  /// Size of world in logical pixels.
  Size worldSize(GridWorld gw) => Size(offset(gw.nCols), offset(gw.nRows));

  /// Where is (square) cell number i?
  double offset(int i) =>
      (lineWidth + (i * (cellWidth + lineWidth))).toDouble();
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:gol_grid/gol_grid.dart';
import 'package:grid_world/grid_world.dart';
import 'package:thumper/thumper.dart';

// ignore_for_file: diagnostic_describe_all_properties
// ignore_for_file: use_key_in_widget_constructors

void main() => runApp(const DemoApp());

/// A GolGrid widget demo.
@immutable
class DemoApp extends StatelessWidget {
  /// Make an app instance.
  const DemoApp({this.isFullScreen = true});

  /// Full screen consumes lots of pixels/CPU - best for Android.
  final bool isFullScreen;

  static const _spacer = SizedBox(height: 3);

  @override
  Widget build(BuildContext context) => MaterialApp(
        title: 'Game of Life Demo',
        theme: ThemeData(
          primarySwatch: Colors.blue,
        ),
        home: Scaffold(
            backgroundColor: Colors.grey,
            appBar: AppBar(
              title: const Text('Game of Life Demo'),
              leading: Icon(Icons.menu),
            ),
            body: isFullScreen ? FullScreenWidget() : _widgetColumn()),
      );

  Widget _widgetColumn() => Center(
        child: ListView(children: <Widget>[
          _spacer,
          GolGrid(DimensionedWorld.make(ConwayEvolver.blinker.lrPadded(6),
              lineWidth: 10, cellWidth: 30)),
          _spacer,
          GolGrid(
              DimensionedWorld.make(ConwayEvolver.pentaDecathlon.clockwise90(),
                  lineWidth: 6, cellWidth: 18),
              foregroundColor: Colors.deepPurple),
          _spacer,
          GolGrid(DimensionedWorld.make(
              ConwayEvolver.lightweightSpaceship
                  .appendBottom(ConwayEvolver.lightweightSpaceship)
                  .appendBottom(ConwayEvolver.lightweightSpaceship)
                  .padRight(50),
              lineWidth: 3,
              cellWidth: 9)),
          _spacer,
          GolGrid(
              DimensionedWorld.make(
                  ConwayEvolver.gliderFleet().padRight(20).padBottom(30),
                  lineWidth: 3,
                  cellWidth: 9),
              foregroundColor: Colors.redAccent),
          _spacer,
          GolGrid(DimensionedWorld.make(
              ConwayEvolver.gosperGliderGun.padded(20),
              lineWidth: 3,
              cellWidth: 9)),
          _spacer,
          GolGrid(
              DimensionedWorld.make(
                  ConwayEvolver.rPentimino.lrPadded(70).tbPadded(60),
                  lineWidth: 1,
                  cellWidth: 3),
              foregroundColor: Colors.white),
          _spacer,
        ]),
      );
}

/// [FullScreenWidget] does a media query to expand a [DimensionedWorld]
/// to fill the media.
///
/// The query only works if this widget is below [MaterialApp] in the tree.
class FullScreenWidget extends StatelessWidget {
  /// A guess as to the media height (virtual pixels) consumed by the AppBar.
  /// TODO: Instead of guessing, look up the appbar via a key, e.g.
  /// medium.com/@diegoveloper/flutter-widget-size-and-position-b0a9ffed9407
  static double get _estimatedAppBarHeight => 80; // 80

  @override
  Widget build(BuildContext context) {
    final s = MediaQuery.of(context).size;
    final dw = DimensionedWorld.make(ConwayEvolver.gunFight()
        .appendBottom(ConwayEvolver.rPentimino.padLeft(30).tbPadded(27))
        .appendBottom(ConwayEvolver.gunFight())
        .lrPadded(6));
    return GolGrid(dw.expandToFit(
        s.width, s.height - _estimatedAppBarHeight - Thumper.height));
  }
}
